framework/src/materials/MaterialData.C

```diff
@@ -88,7 +89,18 @@ MaterialData::eraseProperty(const Elem * elem)
 }
 
 unsigned int
-MaterialData::addPropertyHelper(const std::string & prop_name, const unsigned int state)
+MaterialData::addPropertyHelper(const std::string & prop_name,
+                                const std::type_info & type,
+                                const unsigned int state,
+                                const MaterialBase * const declarer)
 {
-  return _storage.addProperty(prop_name, state);
+  return _storage.addProperty(prop_name, type, state, declarer);
+}
+
+const MaterialBase &
+MaterialData::castRequestorToDeclarer(const MooseObject & requestor) const
+{
+  const auto declarer = dynamic_cast<const MaterialBase *>(&requestor);
+  mooseAssert(declarer, "Not a MaterialBase");
+  return *declarer;
```

Comment:
It could be, but this only ever called for a declarer (which must be a `MaterialBase`):

```
const auto prop_id = addPropertyHelper(
       prop_name, typeid(T), state, declare ? &castRequestorToDeclarer(requestor) : nullptr);
```